### PR TITLE
[SPARK-18066] [CORE] [TESTS] Add Pool usage policies test coverage for FIFO & FAIR Schedulers

### PR DESCRIPTION
**Pull request by erenavsarogullari**
## What changes were proposed in this pull request?

The following FIFO & FAIR Schedulers Pool usage cases need to have unit test coverage :
- FIFO Scheduler just uses **root pool** so even if `spark.scheduler.pool` property is set, related pool is not created and `TaskSetManagers` are added to **root pool**.
- FAIR Scheduler uses `default pool` when `spark.scheduler.pool` property is not set. This can be happened when 
  - `Properties` object is **null**, 
  - `Properties` object is **empty**(`new Properties()`), 
  - **default pool** is set(`spark.scheduler.pool=default`).
- FAIR Scheduler creates a **new pool** with **default values** when `spark.scheduler.pool` property points a **non-existent** pool. This can be happened when **scheduler allocation file** is not set or it does not contain related pool.
## How was this patch tested?

New Unit tests are added.

### CODE REVIEW

**Comment by erenavsarogullari** [comment]
cc @kayousterhout @markhamstra

**Comment by erenavsarogullari** [comment]
Hi @kayousterhout @markhamstra @squito,
This PR aims to extend Pool usage policies unit test coverage and is also ready for review.
Thanks and all feedbacks are welcome in advance ;)

**Comment by squito** [comment]
Jenkins, ok to test

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: I'm all for more tests, but this final case here seems pretty unnecessary

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: minor nit: rename to "FAIR Scheduler creates a new pool when spark.scheduler.pool property points to a non-existent pool"

bigger question:
@markhamstra @kayousterhout Is this really the desired behavior?  Or is there a bug -- should it fail fast?  It looks like [this behavior was intentional](https://github.com/apache/spark/blob/b67b35f76b684c5176dc683e7491fd01b43f4467/core/src/main/scala/org/apache/spark/scheduler/SchedulableBuilder.scala#L143-L145).  Then again, that comment makes it sound like the user is going to directly modify the `weight` and `minShare` of the constructed pool, but though [they are `var`s](https://github.com/apache/spark/blob/b67b35f76b684c5176dc683e7491fd01b43f4467/core/src/main/scala/org/apache/spark/scheduler/Pool.scala#L40-L41), its `private [spark]` and there isn't anything else exposing it to the user.

I prefer fail-fast behavior, but it seems like fair scheduler configuration prefers to assume defaults when there is misconfiguration.  If you think this the right behavior, than there is probably some minor cleanup to do in Pool & SchedulableBuilder to clarify.

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: super nit: the jira doesn't really say anything other than "add these tests", so not much point in listing it here, just name test "FIFO Scheduler just uses root pool".

**Review by squito — COMMENTED** [review]
Thanks for this @erenavsarogullari , great to have more tests.  I have super trivial suggested changes, but one bigger question for other committers I'd like to wait a bit to hear back on.

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: +1. Also how about "FIFO scheduler uses root pool and not spark.scheduler.pool" and then adding a comment at the top of the test that says something like "spark.scheduler.pool should be ignored for the FIFO scheduler, because pools are only needed for fair scheduling"? Took me a minute to figure out what was going on here.

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: Can you re-phrase this as what *should* happen, which makes these comments easier to read? e.g. 

"When task sets are submitted, they should be added to the root pool, and no additional pools should be created (even though there's a configured default pool)."

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: nit: +2 spaces here

**Comment by squito** [comment]
hmm, so I'm thinking about this again, along with your other pr https://github.com/apache/spark/pull/15237

and it got me thinking -- there also aren't any tests for a mix of FIFO and FAIR (at least, I don't see it in "Nested Pool Test" in PoolSuite).  While you are doing this, would you like to add some tests for that as well?  (to be perfectly honest, I need to spend a little bit of time thinking about what that behavior should be anyway, but I will have to do that a little later.)

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: super nit with commenting, but can you eliminate the blank lines on 219 and 221 (which don't help readability) and change this comment to something like

"Submit a new task set manager with pool properties set to null. This should result in the task set manager getting added to the default pool."

(it's helpful in test comments to be clear about what the test is doing itself, versus the expected behavior it's verifying)

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: Similar to the above, can you eliminate the blank lines here, and change this to "When a task set manager is submitted with spark.scheduler.pool unset, it should be added to the default pool (as above)."

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: +1

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: I tracked this down and looks like it was changed [back in 2013](https://github.com/apache/spark/commit/5892393140eb024a32585b6d5b51146ddde8f63a#diff-30ccb25ca46ef204d2446b03ae4f1117R103). @xiajunluan do you remember the motivation for this?

I'm torn because in general I prefer failing fast (e.g., this is better in the case where someone has a small typo they might not notice, and is intending to use a configured pool).  On the other hand, I'm pretty sure I've abused this behavior in the past to avoid configuring pools when I want a bunch of equally weighted pools.  If others are doing that, we shouldn't change the behavior (although I think that, in that case, we should add a big warning / future-deprecation message).

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: can you clean up this comment a bit too?

"The fair scheduler should create a new pool with default values when spark.scheduler.pool points to a pool that doesn't exist yet (this can happen when the file that pools are read from isn't set, or when that file doesn't contain the pool name specified by spark.scheduler.pool)."

**Review by kayousterhout — COMMENTED** [review]
Thanks for adding these tests!

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: yeah I prefer failing fast for the same reason.  But I guess we can't actually change this, since there is a valid use case for this behavior.  We don't really know if others are using this, and I don't think we should change it under them in the next release.

In any case, I still think we should also change `Pool.weight` and `Pool.minShare` to `val`s.

**Comment by kayousterhout** [comment]
@erenavsarogullari what's the status with this PR?  I know you have at least one other fair scheduler PR outstanding, and it makes reviews easier to minimize the number of outstanding things (and helps avoid merge conflicts).

**Comment by erenavsarogullari** [comment]
@kayousterhout thanks for query this PR. I will be submitting this and the other one( #15326) once #16813 is merged.

**Comment by kayousterhout** [comment]
Ok thanks for the update!

**Comment by kayousterhout** [comment]
Ok thanks for the update!

**Comment by erenavsarogullari** [comment]
Jenkins, retest this please

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (erenavsarogullari) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: In this case, at least, i think we can inform the user by emphasizing that the related pool does not exist and it is created with **default settings**. 
Our current logging just points pool creation. WDYT?  

Also Pool `var weight / minShare` have been addressed ;)

**Comment by erenavsarogullari** [comment]
Hi @kayousterhout and @squito,
Firstly, many thanks for the review. All comments are addressed. This is ready to re-review ;)

**Comment by erenavsarogullari** [comment]
build failure looks unrelated (due to timeout at KafkaSourceSuite level)

Jenkins, retest this please

**Comment by erenavsarogullari** [comment]
Jenkins, retest this please

**Comment by erenavsarogullari** [comment]
Hi @kayousterhout and @squito,
This PR is ready to re-review / merge if it is possible.
All feedbacks are welcome and thanks in advance ;)

**Comment by erenavsarogullari** [comment]
Hi @kayousterhout and @squito,
This PR is ready to re-review.
All feedbacks are welcome and thanks in advance ;)

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: JOOC does this result in a warning being printed to the user?  (I know you can't test for that -- but it does seem like useful behavior to have)

**Review by kayousterhout — CHANGES_REQUESTED** [review]
One last small commeent

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (erenavsarogullari) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: Currently, this behavior is **not** highlighted via logs and we just points the pool creation as 
`Created pool: ..., schedulingMode: ..., minShare: ..., weight: ...` so i agree it can be useful for the user and it is addressed with the latest commits.

**Comment by erenavsarogullari** [comment]
Hi @kayousterhout and @squito,
Last comments are addressed so this PR is ready to re-review / merge if it is possible.
Thanks in advance ;)

**Comment by erenavsarogullari** [comment]
The following unit tests look failed due to timeout and _irrelevant_ so triggered again.

**1-** org.apache.spark.storage.BlockManagerProactiveReplicationSuite.proactive block replication - 5 replicas - 4 block manager deletions

**Error Message:** org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 493 times over 5.007521253999999 seconds. Last failure message: 4 did not equal 5.

**2-** org.apache.spark.deploy.SparkSubmitSuite.includes jars passed in through --packages

**Error Message:** org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to failAfter did not complete within 60 seconds.

**Comment by erenavsarogullari** [comment]
Jenkins, retest this please

**Comment by kayousterhout** [comment]
@erenavsarogullari please file a JIRA when you see test failures instead of ignoring them.  I updated https://issues.apache.org/jira/browse/SPARK-19803 for the first failure, but please file a JIRA for the second one.

**Comment by kayousterhout** [comment]
Jenkins retest this please

File: core/src/main/scala/org/apache/spark/scheduler/SchedulableBuilder.scala
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/SchedulableBuilder.scala]: It's not totally clear what the first part of this error message means.  How about changing the message to:

```
logWarning(s"A job was submitted with scheduler pool $poolName, which has not been configured." +
    "This can happen when the file that pools are read from isn't set, or when that file " +
    s"doesn't contain $poolName.  Created $poolName with default configuration (" +
    s"schedulingMode: $DEFAULT_SCHEDULING_MODE, minShare: $DEFAULT_MINIMUM_SHARE, " +
    s"weight: $DEFAULT_WEIGHT)")
```

File: core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/PoolSuite.scala]: Thanks for updating this!

**Review by kayousterhout — COMMENTED** [review]
This looks great -- one last comment update and then I can merge this!

File: core/src/main/scala/org/apache/spark/scheduler/SchedulableBuilder.scala
Comment (erenavsarogullari) [inline core/src/main/scala/org/apache/spark/scheduler/SchedulableBuilder.scala]: Sure, it is updated ;)

**Comment by erenavsarogullari** [comment]
Many thanks @kayousterhout for reviewing this again. Last comment has also been addressed.

Also the following Jira has been created for previous second UT failure.
https://issues.apache.org/jira/browse/SPARK-19964

**Comment by kayousterhout** [comment]
Thanks for your work on this @erenavsarogullari. I've merged this into master.

**Pull request closed without merging**
